cpp/src/arrow/compute/exec/data_holder_node_test.cc

```diff
@@ -0,0 +1,140 @@
+    if (num_batches == 0) {
+      auto empty_record_batch = ExecBatch(*rng_.BatchOf(schema->fields(), 0));
+      out_batches->batches.push_back(empty_record_batch);
+    } else {
+      for (size_t j = 0; j < num_batches; j++) {
+        out_batches->batches.push_back(
+            ExecBatch(*rng_.BatchOf(schema->fields(), batch_size)));
+      }
+    }
+
+    size_t batch_count = out_batches->batches.size();
+    for (int repeat = 1; repeat < multiplicity; ++repeat) {
+      for (size_t i = 0; i < batch_count; ++i) {
+        out_batches->batches.push_back(out_batches->batches[i]);
+      }
+    }
+    out_batches->schema = schema;
+  }
+
+  void CheckRunOutput(const std::vector<BatchesWithSchema>& batches,
+                      const BatchesWithSchema& exp_batches) {
+    ExecContext exec_context(default_memory_pool(),
+                             ::arrow::internal::GetCpuThreadPool());
+
+    ASSERT_OK_AND_ASSIGN(auto plan, ExecPlan::Make(&exec_context));
+
+    Declaration union_decl{"union", ExecNodeOptions{}};
+
+    for (const auto& batch : batches) {
+      union_decl.inputs.emplace_back(Declaration{
+          "source", SourceNodeOptions{batch.schema, batch.gen(/*parallel=*/true,
+                                                              /*slow=*/false)}});
+    }
+    AsyncGenerator<util::optional<ExecBatch>> sink_gen;
+
+    if (batches.size() == 0) {
+      ASSERT_RAISES(Invalid, Declaration::Sequence({union_decl,
+                                                    {"data_holder", ExecNodeOptions{}},
+                                                    {"sink", SinkNodeOptions{&sink_gen}}})
+                                 .AddToPlan(plan.get()));
+      return;
+    } else {
+      ASSERT_OK(Declaration::Sequence({union_decl,
+                                       {"data_holder", ExecNodeOptions{}},
+                                       {"sink", SinkNodeOptions{&sink_gen}}})
+                    .AddToPlan(plan.get()));
+    }
+    Future<std::vector<ExecBatch>> actual = StartAndCollect(plan.get(), sink_gen);
+
+    auto expected_matcher =
+        Finishes(ResultWith(UnorderedElementsAreArray(exp_batches.batches)));
+    ASSERT_THAT(actual, expected_matcher);
+  }
+
+  void CheckDataHolderExecNode(size_t num_input_nodes, size_t num_batches) {
+    auto random_schema = GenerateRandomSchema(num_input_nodes);
+
+    std::vector<std::shared_ptr<RecordBatch>> all_record_batches;
+    std::vector<BatchesWithSchema> input_batches(num_input_nodes);
+    BatchesWithSchema exp_batches;
+    exp_batches.schema = random_schema;
+    for (size_t i = 0; i < num_input_nodes; i++) {
+      GenerateBatchesFromSchema(random_schema, num_batches, &input_batches[i]);
+      for (const auto& batch : input_batches[i].batches) {
+        exp_batches.batches.push_back(batch);
+      }
+    }
+    CheckRunOutput(input_batches, exp_batches);
+  }
+
+  ::arrow::random::RandomArrayGenerator rng_;
+};
+
+TEST_F(TestDataHolderNode, TestNonEmpty) {
```

Comment:
This will eventually need tests that cover the case where limits are hit and batches are actually persisted.